### PR TITLE
Add serializer interface

### LINKED ISSUE #331: Add serializer interface
Is there any reason a serializer interface does not exist? An interface type hint in `PersistentQueue` / `PersistentFactory` especially would make things a little more flexible.

### PR DESCRIPTION
**Pull request by namoscato**
Resolves #331

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -3,3 +3,4 @@ vendor
 phpunit.xml
 coverage
 _build
+.idea/
```
Comment (acrobat) [inline .gitignore]: This should be defined in your global gitignore. It's not related to this project specific

**Comment by namoscato** [comment]
I forgot to swap out the type hint in `PersistentFactory` as well. Thanks for the review, @acrobat!

**Comment by sagikazarmark** [comment]
Thanks for the contribution @namoscato 

Can you please give me some reasoning why this change brings value to the project?

The serializer is nothing, but a facade around a number of things which in turn rely on the Symfony Serializer/Deserializer interfaces.

One can argue about the complexity of this setup, but introducing an interface on this level would be somewhat redundant in my opinion. There is nothing you can't do with the current architecture.

So thanks again, but at this time I'm against merging this (still would love to hear any reasoning you can give).

**Comment by namoscato** [comment]
@sagikazarmark, yeah no problem.

I am using this project with an application that relies solely on [`jms/serializer`](http://jmsyst.com/libs/serializer) for serialization. I am currently extending the concrete `Bernard\Serializer` and effectively overriding the entire implementation so that I can still use `PersistentQueue` / `PersistentFactory`. This works, but it just seems a bit awkward.

**Comment by sagikazarmark** [comment]
I think a better solution would be providing some kind of bridge using the Symfony interfaces. That way you wouldn't have to care about serializing the envelope since it's already handled by the aggregate.

**Comment by namoscato** [comment]
Yeah, that would be nice. Unfortunately, `jms/serializer` does not clearly split its serialization into normalization and encoding phases per Symfony's pattern; thus, I did not see a straightforward way to slip it into the aggregate normalization.

**Comment by sagikazarmark** [comment]
You don't necessarily have to separate them, usually the two interfaces are implemented on one class.

But I haven't done that before, so it might not be that easy as I imagine.

**Comment by namoscato** [comment]
Yeah, you know what, I am actually using a pretty outdated version of `jms/serializer`, and it seems they have since introduced an [`ArrayTransformerInterface`](https://github.com/schmittjoh/serializer/blob/ddb7d0df42bc8054842068276f48f33bf75d0acb/src/JMS/Serializer/ArrayTransformerInterface.php) which seems to address the problem I was having.

Thanks for the feedback, @sagikazarmark!

**Pull request closed without merging**